Makes sure a SuperAdministrator can publish on any sites of the network.

imath opened the pull request:
## Description

After preloading the current user from the `/wp/v2/users/me` REST request, a Super Administrator on a site he's not a user of has no caps/roles on it, as a result the major action button is set to `Submit for review`. The changes are making sure a Super Administrator with no caps/roles on the site can publish a post from  the Gutenberg editor just like it's the case within the Classic Editor.

### Steps to reproduce the bug on a multisite configuration

1. Log in as a SuperAdmin
2. Create a user
3. Create a site for this user.
4. Try to add a post on the created site.

## How Has This Been Tested?
I've tested this on a Multisite configuration of WordPress version 5.0-alpha-42125-src
My testing environment: regular macOS HighSierra AMP - PHP 7.1.7
I've checked changes had no effects on the contributor role of the site.

## Screenshots (jpeg or gifs if applicable):

### Without the patch

![without-patch](https://user-images.githubusercontent.com/1834524/33800368-ff40e23c-dd3e-11e7-91e3-bbaa76ecd12e.png)

### With the patch

![with-patch](https://user-images.githubusercontent.com/1834524/33800392-50746fa2-dd3f-11e7-968e-9d5d137e8286.png)


## Types of changes

Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

imath committed:
Makes sure a SuperAdministrator can publish on any sites of the network.

imath committed:
Makes sure a SuperAdministrator can publish on any sites of the network.

youknowriad commented:
Thanks for this fix @imath I wonder if this could be considered a REST API bug.

Could this fix #2702 cc @aduth

imath commented:
You're welcome @youknowriad I'm going to improve this PR because, now i've read #2702 i don't think it's the right way to do it.

First i've hesitated about the fact it might be related to the REST API. Now my ideas are clearer, i'm pretty sure it's not.

I think Gutenberg shouldn't rely on the values that are stored into the database for capabilities. The reason is you can have plugins containing capability mapping. And you won't have this information into the wp_{blog_id}_capabilities meta field. The REST API basically returns the value of this meta field.

A better way to handle this imho is to get the post type object capabilities for the post type being edited  and to loop in it for the current user checking `current_user_can()` to be sure to trigger the `map_meta_cap` filter and avoid the super admin thing i've found in this PR and other troubles that may be the origin of #2702.

If you look at the `post_submit_meta_box()`, you'll see at the beginning that Core uses current_user_can() at line 27 of `wp-admin/includes/meta-boxes.php`

I'll update the PR in this direction.

aduth commented:
>I think Gutenberg shouldn't rely on the values that are stored into the database for capabilities.

Wouldn't this mean that, more broadly than just Gutenberg, the `capabilities` field on the REST API user entity cannot be trusted?

aduth commented:
Enumerated at https://github.com/WordPress/gutenberg/issues/2702#issuecomment-330696191, there are quite a few places we test for capabilities to determine whether something should be rendered.

imath committed:
Merge branch 'master' into fix/superadmin-caps

Update branch to upstream master

imath commented:
> Wouldn't this mean that, more broadly than just Gutenberg, the capabilities field on the REST API user entity cannot be trusted?

To be sure a user can do a specific cap, you need to use `current_user_can()` so that the map_meta_cap filter is fired & the Super Admin check is performed.

Saving a cap for a specific user is not always done, so yes there's a good chance (and it's more likely to be the case when dealing with custom post types as they can set custom caps / caps mapping) the meta field does not reflect precisely what a user can do.

About the User controller of the REST API, i'm not sure its goal is to get what a user can do with a specific post type, but more to get/create/update/delete a user. If it was the case, the post type name would have been one of the available query arguments i think.

imath committed:
Validate the User's capabilities for the Post Type being edited

imath committed:
Validate the User's capabilities for the Post Type being edited

imath committed:
Add missing full-stops to inline comments

imath committed:
Add missing full-stops to inline comments

imath commented:
@youknowriad & @aduth 

I've updated the PR the way i've explained in [this comment](#issuecomment-350725508). It's fixing the issue i've first met with a SuperAdmin with no role on a child site, it should also fix possible issues for Custom Post Types having custom capabilities.

I think it shouldn't need any changes on the JavaScript part, as i kind of map back custom capabilities to post capabilities.

youknowriad commented on lib/client-assets.php:
```diff
@@ -208,6 +208,40 @@ function gutenberg_preload_api_request( $memo, $path ) {
+
+				$response->set_data(
+					array(
+						'capabilities' => $post_type_caps,
```
Really awesome work here Thanks.

As far as I understand, these capabilities are "post type" specific capabilities only. Do you think it makes more sense to use a more descriptive key `post_type_capabilties => array( $post_type => $post_type_caps )` to avoid confusion on the client-side?

Granted this means we need to update some client code as well.

Also this raises the question of whether this should be a separate endpoint `/users/me/post_type_caps?postType=post`

imath replied:
You're very welcome @youknowriad 

I agree it makes a lot of sense. As I lack some knowledge of React, I took no risk adventuring myself into this part of Gutenberg's code :smiley:

As you said, it would require some updates into the client code so that the`post_type_capabilties` are used if available.

About the new endpoint, yes it's a possibility. You can alternatively use a specific query parameter `/users/me/?context=edit&post_type=post` and temporarly use the `rest_prepare_user` filter to include the `post_type_capabilties` into the response before editing the `WP_REST_Users_Controller::prepare_item_for_response()` method once Gutenberg is merged into core.

youknowriad replied:
> About the new endpoint, yes it's a possibility. You can alternatively use a specific query parameter /users/me/?context=edit&post_type=post and temporarly use the rest_prepare_user filter to include the post_type_capabilties into the response before editing the WP_REST_Users_Controller::prepare_item_for_response() method once Gutenberg is merged into core.

That sounds like a good solution for now. We'll probably have to bring this issue to the REST API team later. Might be good to have @aduth's insights

youknowriad replied:
Also, the client changes should be small enough, I can help with those if necessary ;)

imath committed:
Merge branch 'master' into fix/superadmin-caps

Update branch to latest upstream edits.

imath committed:
Add a 'post_type_capabilities' custom field to the REST API response of user.

Edit the query parameters used for the users/me endpoint on the server and client side.
Merge capabilities with post type capabilities into the client side before checking the user capabilities

imath committed:
Add a 'post_type_capabilities' custom field to the REST API response of user.

Edit the query parameters used for the users/me endpoint on the server and client side.
Merge capabilities with post type capabilities into the client side before checking the user capabilities

imath committed:
Fix failing travis tests

Add spaces into inline comments.
Add missing "gutenberg" textdomain.
Remove quotes in JS object keys.

imath committed:
Fix failing travis tests

Add spaces into inline comments.
Add missing "gutenberg" textdomain.
Remove quotes in JS object keys.

imath commented on editor/components/post-publish-button/label.js:
```diff
@@ -28,7 +28,10 @@ export function PublishButtonLabel( {
+	const userCaps = user.data ?
+		{ ...user.data.capabilities, ...user.data.post_type_capabilities } :
+		{ publish_posts: false };
+	const isContributor = ! userCaps.publish_posts;
```
Note to self: the test is probably failing because i don't test user.data anymore when setting isContributor.

youknowriad commented on editor/components/post-author/check.js:
```diff
@@ -11,17 +11,23 @@ import { compose } from '@wordpress/element';
-	if ( ! user.data || ! user.data.capabilities.publish_posts || authors.length < 2 ) {
+	const userCaps = user.data ?
+		{ ...user.data.capabilities, ...user.data.post_type_capabilities } :
+		{ 'publish_posts': false };
```
Do we need if this will be always a "post type capability" or not? In which case we coud avoid using `user.data.capabilities` entirely?

Minor: maybe we could avoid creating a separate object, by leveraging lodash `get`

`const canPublishPosts = get( user, 'data.post_type_capabilities.publish_posts',  false )`

imath replied:
Good point, i think we don't need it, i'll update the PR in this direction + thanks for the tip about lodash :+1:

youknowriad commented on editor/components/post-author/check.js:
```diff
@@ -11,17 +11,23 @@ import { compose } from '@wordpress/element';
 }
 
 const applyWithAPIData = withAPIData( () => {
+	const postTypeSlug = window._wpGutenbergPost.type;
```
We should avoid using the global here, and rely on the `getCurrentPostType` selector. Example here 
https://github.com/WordPress/gutenberg/blob/master/editor/components/page-attributes/check.js#L32-L51

imath replied:
Thanks for your feedback. I've tried to use `getCurrentPostType` but unfortunately the Post Type was not defined yet at the time the request was sent. I'll give it another look asap.

imath committed:
Stop using a global to get the post type name

imath committed:
Stop using a global to get the post type name

imath commented:
@youknowriad 

I improved the PR the way you advised (use lodash/ stop using the global to get the post type / only use post type capabilities).

I still need to update the tests so that they also use the `post_type_capabilities`. I think it's the reason why many failed during my last update.

imath committed:
Update tests so that they use post_type_capabilities

imath committed:
Update tests so that they use post_type_capabilities

imath commented:
I confirm. I just updated tests so that they use `post_type_capabilities` instead of `capabilites` and they are now passing.

imath committed:
Merge branch 'master' into fix/superadmin-caps

imath commented:
Hi @youknowriad @aduth 

As 3 days passed and this issue hasn't been updated, I guess i can't do anything more on my side. So I'm leaving you deal with it, hoping the Gutenberg UI will soon be consistent with the Post Type capabilities of a user.

Thanks for your consideration and advices.

youknowriad commented:
Sorry for the delay @imath I was AFK, this looks solid to me. Might be good to have a second opinion @aduth before merging?

youknowriad committed:
Data Module: Restrict the state access to the module registering the reducer only

youknowriad committed:
Blocks: Add the missing alignments classNames to the Cover Image Block's markup

aduth reviewed (commented):
Speaking to the REST API changes specifically, the `post_type_capabilities` property _appears_ at first glance redundant with the existing `capabilities` property. It's a bit strange to specify a post type parameter on a user endpoint, though more understandable with the endpoint returning capabilities information.

Curious though, would another option here have been to add an `is_super_admin` property to the endpoint to allow all actions?

https://developer.wordpress.org/reference/functions/is_super_admin/

aduth commented:
Or, alternatively, should we be running the `capabilities` property on the users endpoint through relevant filters rather than sending raw?

imath commented on lib/client-assets.php:
```diff
@@ -708,9 +708,12 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 	// Preload common data.
 	$preload_paths = array(
-		'/wp/v2/users/me?context=edit',
+		sprintf( '/wp/v2/users/me?post_type=%s&context=edit', sanitize_key( $post_type ) ),
```
Sorry, it's probably my habit to always sanitize when i use sprintf. I'll remove it.

aduth replied:
>Sorry, it's probably my habit to always sanitize when i use sprintf. I'll remove it.

I think sanitization is sensible, just making sure that it aligns with what's permitted when registering a custom post type.

imath commented:
@aduth 

> Curious though, would another option here have been to add an is_super_admin property to the endpoint to allow all actions?

I should probably edit the Title of the PR. It does concerned any user having caps to edit a post type not only the Super Administrator.

You can have specific roles in WordPress to edit/publish etc.. a specific custom post type.

> Or, alternatively, should we be running the capabilities property on the users endpoint through relevant filters rather than sending raw?

Sorry i am not sure to understand. Could you explain in a more detailed way ?

aduth commented:
>Sorry i am not sure to understand. Could you explain in a more detailed way ?

I mean instead of naively returning `$user->allcaps` here:

https://github.com/WordPress/wordpress-develop/blob/bc35d72a6363fbf67f65bc96f22fd758e9495e2a/src/wp-includes/rest-api/endpoints/class-wp-rest-users-controller.php#L893-L895

...that we'd iterate through known capabilities and call `user_can` to ensure that the relevant filters are applied†. Though, in the case of a super admin, I expect `allcaps` would be an incomplete array of capabilities, since they're not explicitly assigned for the user.

_† Might be a non-negligible performance impact to this._

youknowriad commented:
> .Might be a non-negligible performance impact to this.

Also, the capabilities may change per CPT if I understand properly. So we'll have to loop over the CPTs and enforce a per CPT structure in this case.

imath commented:
@aduth @youknowriad 

Thanks a  lot for the detailed explanation. It's an interesting point. I don't know about performance. But it might add some complexity to the response.

Post Type [capabilities](https://developer.wordpress.org/reference/functions/get_post_type_capabilities/) can be the same than the regular `post` post type or can be specific. This is defined when registering the post type. These caps are inside the WordPress global `$wp_post_types`.

In the editor's components, the checked capabilities are:
- `publish_posts`
- `edit_others_posts` (for the sticky feature)

I think `upload_files` should be also checked to eventually disable image/audio/video blocks but that's another issue #3672

If you get all post type objects using `get_post_types( array( 'show_in_rest' => true ), 'objects' )` and loop into each object, you'll need to build a more complex array to keep on using the above  'generic' capabilities into Gutenberg components, for instance:

```php
array( 'post_types_caps' =>
	array(
		'post' => array(
			'publish_posts' => current_user_can( 'publish_posts' ),
		),
		'book' => array(
			'publish_posts' => current_user_can( 'publish_books' ),
		),
	),
);
```

otherwise the 'publish_posts' cap will be the one of the last `publish_{post_type_cap}`. Then in components, you'd also need to get the current post type into the functions that render elements
eg:
`const isContributor = user.data && ! user.data.post_type_caps[postTypeSlug].publish_posts;`

So i'd say why not, but i think what suggested @youknowriad to use a different key for capabilities about the current post type (`post_type_capabilities`) clearly separate it from the regular capabilities to avoid confusion, keeps things simple and fixes the issue :)

imath committed:
Merge branch 'master' into fix/superadmin-caps

imath committed:
Apply @aduth recommandations

Call get using an array to improve performance.
Sanitizing the post type slug is not necessary

imath committed:
Apply @aduth recommandations

Call get using an array to improve performance.
Sanitizing the post type slug is not necessary

aduth committed:
Factory: Fix block creation with falsey default

karmatosed committed:
3 column inserter

This is a suggestion for #4080. It increases the 2 columns to 3 for the inserter.

imath committed:
Update path to selectors as it has been edited upstream

Commit:
Try improving mobile controls

Those controls include the mover and the ellipsis menu. As such, this PR fixes #4084.

What this PR does, is remove all mobile tools from the quick toolbar, and instead shows these tools inline at the bottom of a _selected_ block.

Commit:
Improve paragraph block description.

Props @maddisondesigns for text.

Commit:
Rename "Classic Text" to "Classic".

Fixes #2956.

This renames the "Classic Text" block to just "Classic", because it can do more than just text.

karmatosed committed:
Widens the inserter

Adds in a wider inserter.

karmatosed committed:
Left aligns the title

karmatosed committed:
Move title to center again

The left align wasn't working so lets move border and center align.

Commit:
Remove references to obsolete class.

youknowriad committed:
Block Mobile Toolbar: Extract as a separate component

youknowriad committed:
Merge branch 'try/improved-mobile-tools' of github.com:WordPress/gutenberg into try/improved-mobile-tools

ryo-utsunomiya committed:
Define setAttributes before using it.

jasmussen committed:
Merge pull request #4103 from WordPress/update/improve-paragraph-description

Improve paragraph block description.

jasmussen committed:
Merge pull request #4101 from WordPress/try/improved-mobile-tools

Try improving mobile controls

youknowriad commented:
Since @aduth is AFK, I don't want to let this PR stale. In opinion, this is good enough to ship. It's certainly not idea to add this `post_type_capabilities` attribute but it's also not idea to iterate over all the post types and compute the capabilities for each one of them.

A separate endpoint `/user/me/post_type_catabilities` might make sense here but I'm ok with moving forward and revisit. This fixes several bugs in the multisite installations.

I'd appreciate another review, maybe @mcsf @pento @gziolo  before merging.

Also, this needs a rebase :)

mcsf commented on lib/register.php:
```diff
@@ -451,6 +451,66 @@ function gutenberg_register_rest_api_post_revisions() {
+ * @param WP_REST_Request $request Full details about the REST API request.
+ * @return object The Post Type capabilities.
+ */
+function gutenberg_get_post_type_capabities( $user, $name, $request ) {
```
Typo at `capabities`

mcsf commented on lib/register.php:
```diff
@@ -451,6 +451,66 @@ function gutenberg_register_rest_api_post_revisions() {
+/**
+ * Includes the value for the custom field `post_type_capabities` inside the REST API response of user.
+ *
+ * TODO: This is a temporary solution. Next step would be to edit theWP_REST_Users_Controller,
```
Missing space at `theWP_REST_Users_Controller`.

Wondering if we should be grouping these things that are expected to be fixed directly in core. Until then, does this function, as well as the next, make more sense in `lib/compat.php`? It might make them more visible.

mcsf reviewed (commented):
This is a nice fix; thank you, @imath. As mentioned, this should be rebased.

There's room to smooth the edges (e.g. refactor the common patterns), but we can address that later.

jorgefilipecosta committed:
Removed duplicate of buildTermsTree in HierarchicalTermSelector. (#4111)

Commit:
Merge pull request #4104 from WordPress/update/rename-classic-text

Rename "Classic Text" to "Classic". Props @jasmussen

Commit:
Merge pull request #4095 from WordPress/try/three-column-inserter

3 column inserter

imath committed:
Apply @mcsf recommandations

Fix a missing space in two docblocks
Fix a typo in a function name
Move temporary functions to lib/compat.php to improve their visibility

imath committed:
Fix some conflicts which appeared after rebasing

imath committed:
Remove new functions from lib/register.php

These functions are now in lib/compat.php to follow @mcsf recommandations.

imath commented:
@mcsf Hi,

yw & thanks a lot for your recommandations, i've applied them and "rebased". As it was the first time i was using this command, i hope i didn't put a mess 😰

youknowriad commented:
Mmmm! the rebase look broken :( Might be simpler to copy the code and push another PR? You can probably try cherry-picking the right commits as well?

Do you think you can open another PR to fix this?

imath commented:
@youknowriad i was fearing it, so i saved a diff before 😉 it will be easier for me to create another PR, i think i still need to train to be a good rebaser! I’ll do it tonight (office hours for me!) and i’ll add it to this PR. Sorry for the « contretemps » 😊

youknowriad commented:
No worries, Git can be scary sometimes :)

youknowriad commented:
superseded by #4122
